### PR TITLE
remove value object boilerplate

### PR DESCRIPTION
**Pull request by sgatchell**
Used [goog's ValueObject](https://github.com/google/auto/blob/master/value/userguide/builders.md) to remove lots of boilerplate in this file.  The generated code is below
``` java

package org.ambraproject.rhino.model.article;

import com.google.common.collect.ImmutableList;
import java.time.LocalDate;
import java.util.List;
import javax.annotation.Generated;
import javax.annotation.Nullable;

@Generated("com.google.auto.value.processor.AutoValueProcessor")
 final class AutoValue_ArticleMetadata extends ArticleMetadata {

  private final String doi;
  private final String title;
  private final String eissn;
  private final String journalName;
  private final String description;
  private final String rights;
  private final String language;
  private final String format;
  private final Integer pageCount;
  private final String eLocationId;
  private final LocalDate publicationDate;
  private final String volume;
  private final String issue;
  private final String publisherLocation;
  private final String publisherName;
  private final String url;
  private final String nlmArticleType;
  private final String articleType;
  private final ImmutableList<AssetMetadata> assets;
  private final ImmutableList<RelatedArticleLink> relatedArticles;
  private final ImmutableList<NlmPerson> editors;

  private AutoValue_ArticleMetadata(
      @Nullable String doi,
      @Nullable String title,
      @Nullable String eissn,
      @Nullable String journalName,
      @Nullable String description,
      @Nullable String rights,
      @Nullable String language,
      @Nullable String format,
      @Nullable Integer pageCount,
      @Nullable String eLocationId,
      @Nullable LocalDate publicationDate,
      @Nullable String volume,
      @Nullable String issue,
      @Nullable String publisherLocation,
      @Nullable String publisherName,
      @Nullable String url,
      @Nullable String nlmArticleType,
      @Nullable String articleType,
      @Nullable ImmutableList<AssetMetadata> assets,
      @Nullable ImmutableList<RelatedArticleLink> relatedArticles,
      @Nullable ImmutableList<NlmPerson> editors) {
    this.doi = doi;
    this.title = title;
    this.eissn = eissn;
    this.journalName = journalName;
    this.description = description;
    this.rights = rights;
    this.language = language;
    this.format = format;
    this.pageCount = pageCount;
    this.eLocationId = eLocationId;
    this.publicationDate = publicationDate;
    this.volume = volume;
    this.issue = issue;
    this.publisherLocation = publisherLocation;
    this.publisherName = publisherName;
    this.url = url;
    this.nlmArticleType = nlmArticleType;
    this.articleType = articleType;
    this.assets = assets;
    this.relatedArticles = relatedArticles;
    this.editors = editors;
  }

  @Nullable
  @Override
  public String getDoi() {
    return doi;
  }

  @Nullable
  @Override
  public String getTitle() {
    return title;
  }

  @Nullable
  @Override
  public String getEissn() {
    return eissn;
  }

  @Nullable
  @Override
  public String getJournalName() {
    return journalName;
  }

  @Nullable
  @Override
  public String getDescription() {
    return description;
  }

  @Nullable
  @Override
  public String getRights() {
    return rights;
  }

  @Nullable
  @Override
  public String getLanguage() {
    return language;
  }

  @Nullable
  @Override
  public String getFormat() {
    return format;
  }

  @Nullable
  @Override
  public Integer getPageCount() {
    return pageCount;
  }

  @Nullable
  @Override
  public String geteLocationId() {
    return eLocationId;
  }

  @Nullable
  @Override
  public LocalDate getPublicationDate() {
    return publicationDate;
  }

  @Nullable
  @Override
  public String getVolume() {
    return volume;
  }

  @Nullable
  @Override
  public String getIssue() {
    return issue;
  }

  @Nullable
  @Override
  public String getPublisherLocation() {
    return publisherLocation;
  }

  @Nullable
  @Override
  public String getPublisherName() {
    return publisherName;
  }

  @Nullable
  @Override
  public String getUrl() {
    return url;
  }

  @Nullable
  @Override
  public String getNlmArticleType() {
    return nlmArticleType;
  }

  @Nullable
  @Override
  public String getArticleType() {
    return articleType;
  }

  @Nullable
  @Override
  public ImmutableList<AssetMetadata> getAssets() {
    return assets;
  }

  @Nullable
  @Override
  public ImmutableList<RelatedArticleLink> getRelatedArticles() {
    return relatedArticles;
  }

  @Nullable
  @Override
  public ImmutableList<NlmPerson> getEditors() {
    return editors;
  }

  @Override
  public String toString() {
    return "ArticleMetadata{"
        + "doi=" + doi + ", "
        + "title=" + title + ", "
        + "eissn=" + eissn + ", "
        + "journalName=" + journalName + ", "
        + "description=" + description + ", "
        + "rights=" + rights + ", "
        + "language=" + language + ", "
        + "format=" + format + ", "
        + "pageCount=" + pageCount + ", "
        + "eLocationId=" + eLocationId + ", "
        + "publicationDate=" + publicationDate + ", "
        + "volume=" + volume + ", "
        + "issue=" + issue + ", "
        + "publisherLocation=" + publisherLocation + ", "
        + "publisherName=" + publisherName + ", "
        + "url=" + url + ", "
        + "nlmArticleType=" + nlmArticleType + ", "
        + "articleType=" + articleType + ", "
        + "assets=" + assets + ", "
        + "relatedArticles=" + relatedArticles + ", "
        + "editors=" + editors
        + "}";
  }

  @Override
  public boolean equals(Object o) {
    if (o == this) {
      return true;
    }
    if (o instanceof ArticleMetadata) {
      ArticleMetadata that = (ArticleMetadata) o;
      return ((this.doi == null) ? (that.getDoi() == null) : this.doi.equals(that.getDoi()))
           && ((this.title == null) ? (that.getTitle() == null) : this.title.equals(that.getTitle()))
           && ((this.eissn == null) ? (that.getEissn() == null) : this.eissn.equals(that.getEissn()))
           && ((this.journalName == null) ? (that.getJournalName() == null) : this.journalName.equals(that.getJournalName()))
           && ((this.description == null) ? (that.getDescription() == null) : this.description.equals(that.getDescription()))
           && ((this.rights == null) ? (that.getRights() == null) : this.rights.equals(that.getRights()))
           && ((this.language == null) ? (that.getLanguage() == null) : this.language.equals(that.getLanguage()))
           && ((this.format == null) ? (that.getFormat() == null) : this.format.equals(that.getFormat()))
           && ((this.pageCount == null) ? (that.getPageCount() == null) : this.pageCount.equals(that.getPageCount()))
           && ((this.eLocationId == null) ? (that.geteLocationId() == null) : this.eLocationId.equals(that.geteLocationId()))
           && ((this.publicationDate == null) ? (that.getPublicationDate() == null) : this.publicationDate.equals(that.getPublicationDate()))
           && ((this.volume == null) ? (that.getVolume() == null) : this.volume.equals(that.getVolume()))
           && ((this.issue == null) ? (that.getIssue() == null) : this.issue.equals(that.getIssue()))
           && ((this.publisherLocation == null) ? (that.getPublisherLocation() == null) : this.publisherLocation.equals(that.getPublisherLocation()))
           && ((this.publisherName == null) ? (that.getPublisherName() == null) : this.publisherName.equals(that.getPublisherName()))
           && ((this.url == null) ? (that.getUrl() == null) : this.url.equals(that.getUrl()))
           && ((this.nlmArticleType == null) ? (that.getNlmArticleType() == null) : this.nlmArticleType.equals(that.getNlmArticleType()))
           && ((this.articleType == null) ? (that.getArticleType() == null) : this.articleType.equals(that.getArticleType()))
           && ((this.assets == null) ? (that.getAssets() == null) : this.assets.equals(that.getAssets()))
           && ((this.relatedArticles == null) ? (that.getRelatedArticles() == null) : this.relatedArticles.equals(that.getRelatedArticles()))
           && ((this.editors == null) ? (that.getEditors() == null) : this.editors.equals(that.getEditors()));
    }
    return false;
  }

  @Override
  public int hashCode() {
    int h = 1;
    h *= 1000003;
    h ^= (doi == null) ? 0 : this.doi.hashCode();
    h *= 1000003;
    h ^= (title == null) ? 0 : this.title.hashCode();
    h *= 1000003;
    h ^= (eissn == null) ? 0 : this.eissn.hashCode();
    h *= 1000003;
    h ^= (journalName == null) ? 0 : this.journalName.hashCode();
    h *= 1000003;
    h ^= (description == null) ? 0 : this.description.hashCode();
    h *= 1000003;
    h ^= (rights == null) ? 0 : this.rights.hashCode();
    h *= 1000003;
    h ^= (language == null) ? 0 : this.language.hashCode();
    h *= 1000003;
    h ^= (format == null) ? 0 : this.format.hashCode();
    h *= 1000003;
    h ^= (pageCount == null) ? 0 : this.pageCount.hashCode();
    h *= 1000003;
    h ^= (eLocationId == null) ? 0 : this.eLocationId.hashCode();
    h *= 1000003;
    h ^= (publicationDate == null) ? 0 : this.publicationDate.hashCode();
    h *= 1000003;
    h ^= (volume == null) ? 0 : this.volume.hashCode();
    h *= 1000003;
    h ^= (issue == null) ? 0 : this.issue.hashCode();
    h *= 1000003;
    h ^= (publisherLocation == null) ? 0 : this.publisherLocation.hashCode();
    h *= 1000003;
    h ^= (publisherName == null) ? 0 : this.publisherName.hashCode();
    h *= 1000003;
    h ^= (url == null) ? 0 : this.url.hashCode();
    h *= 1000003;
    h ^= (nlmArticleType == null) ? 0 : this.nlmArticleType.hashCode();
    h *= 1000003;
    h ^= (articleType == null) ? 0 : this.articleType.hashCode();
    h *= 1000003;
    h ^= (assets == null) ? 0 : this.assets.hashCode();
    h *= 1000003;
    h ^= (relatedArticles == null) ? 0 : this.relatedArticles.hashCode();
    h *= 1000003;
    h ^= (editors == null) ? 0 : this.editors.hashCode();
    return h;
  }

  static final class Builder extends ArticleMetadata.Builder {
    private String doi;
    private String title;
    private String eissn;
    private String journalName;
    private String description;
    private String rights;
    private String language;
    private String format;
    private Integer pageCount;
    private String eLocationId;
    private LocalDate publicationDate;
    private String volume;
    private String issue;
    private String publisherLocation;
    private String publisherName;
    private String url;
    private String nlmArticleType;
    private String articleType;
    private ImmutableList<AssetMetadata> assets;
    private ImmutableList<RelatedArticleLink> relatedArticles;
    private ImmutableList<NlmPerson> editors;
    Builder() {
    }
    @Override
    public ArticleMetadata.Builder setDoi(@Nullable String doi) {
      this.doi = doi;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setTitle(@Nullable String title) {
      this.title = title;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setEissn(@Nullable String eissn) {
      this.eissn = eissn;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setJournalName(@Nullable String journalName) {
      this.journalName = journalName;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setDescription(@Nullable String description) {
      this.description = description;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setRights(@Nullable String rights) {
      this.rights = rights;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setLanguage(@Nullable String language) {
      this.language = language;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setFormat(@Nullable String format) {
      this.format = format;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setPageCount(@Nullable Integer pageCount) {
      this.pageCount = pageCount;
      return this;
    }
    @Override
    public ArticleMetadata.Builder seteLocationId(@Nullable String eLocationId) {
      this.eLocationId = eLocationId;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setPublicationDate(@Nullable LocalDate publicationDate) {
      this.publicationDate = publicationDate;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setVolume(@Nullable String volume) {
      this.volume = volume;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setIssue(@Nullable String issue) {
      this.issue = issue;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setPublisherLocation(@Nullable String publisherLocation) {
      this.publisherLocation = publisherLocation;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setPublisherName(@Nullable String publisherName) {
      this.publisherName = publisherName;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setUrl(@Nullable String url) {
      this.url = url;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setNlmArticleType(@Nullable String nlmArticleType) {
      this.nlmArticleType = nlmArticleType;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setArticleType(@Nullable String articleType) {
      this.articleType = articleType;
      return this;
    }
    @Override
    public ArticleMetadata.Builder setAssets(@Nullable List<AssetMetadata> assets) {
      this.assets = (assets == null ? null : ImmutableList.copyOf(assets));
      return this;
    }
    @Override
    public ArticleMetadata.Builder setRelatedArticles(@Nullable List<RelatedArticleLink> relatedArticles) {
      this.relatedArticles = (relatedArticles == null ? null : ImmutableList.copyOf(relatedArticles));
      return this;
    }
    @Override
    public ArticleMetadata.Builder setEditors(@Nullable List<NlmPerson> editors) {
      this.editors = (editors == null ? null : ImmutableList.copyOf(editors));
      return this;
    }
    @Override
    public ArticleMetadata build() {
      return new AutoValue_ArticleMetadata(
          this.doi,
          this.title,
          this.eissn,
          this.journalName,
          this.description,
          this.rights,
          this.language,
          this.format,
          this.pageCount,
          this.eLocationId,
          this.publicationDate,
          this.volume,
          this.issue,
          this.publisherLocation,
          this.publisherName,
          this.url,
          this.nlmArticleType,
          this.articleType,
          this.assets,
          this.relatedArticles,
          this.editors);
    }
  }

}
```

### CODE REVIEW

**Comment by sgatchell** [comment]
@sjtower This is something I did just to get back into the java game.  I don't expect this to be merged, it's just an experiment. Just wanted to solicit your opinion.

File: src/main/java/org/ambraproject/rhino/model/article/ArticleMetadata.java [hunk 1]
```diff
@@ -23,358 +23,74 @@
...
+  }
+  
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract ArticleMetadata build();
+    
+    public abstract Builder setDoi(String doi);
+    public abstract Builder setTitle(String title);
+    public abstract Builder setEissn(String eIssn);
+    // journalName was not previously chainable.  Is that a problem?
```
Comment (sjtower) [inline src/main/java/org/ambraproject/rhino/model/article/ArticleMetadata.java]: no problem :)

**Comment by sjtower** [comment]
removed ~250 lines of code with a single annotation? sounds like a great plan to me. Feel free to make a ticket for this change so it can be properly merged.

**Comment by sgatchell** [comment]
Will do, thanks!

**Comment by egh** [comment]
Wow.

**Comment by sjtower** [comment]
:+1: 
please include ticket numbers in your branch names in the future, EG:

dev-RARO-856

**Pull request merged**